### PR TITLE
USB: esp_tinyusb - fix memory leak in case of error, refactoring

### PR DESCRIPTION
**Pull request by esp-saurabhbansal**
# Checklist

- [ ] Component contains License
- [ ] Component contains README.md
- [ ] Component contains idf_component.yml file with `url` field defined
- [ ] Component was added to [upload job](https://github.com/espressif/idf-extra-components/blob/master/.github/workflows/upload_component.yml#L18)
- [ ] Component was added to [build job](https://github.com/espressif/idf-extra-components/blob/master/test_app/CMakeLists.txt#L8)
- [ ] _Optional:_ Component contains unit tests
- [ ] CI passing

# Change description
1. Fix memory leak while returning in case of error. While returning in case of error, these should be called for freeing up the resources:
    a. tinyusb_cdc_deinit()
    b. free(acm->rx_tfbuf);
2. Remove unused 'initialized'
3. Remove hard coding of "/dev/tusb_cdc"

### CODE REVIEW

**Comment by esp-saurabhbansal** [comment]
@tore-espressif @roma-jam  @igrr 
PTAL

File: usb/esp_tinyusb/tusb_cdc_acm.c
Comment (tore-espressif) [inline usb/esp_tinyusb/tusb_cdc_acm.c:387]: Nitpick: Could we use the `ESP_GOTO_ON_FALSE` macros here? to have the same style as in IDF code?

File: usb/esp_tinyusb/tusb_cdc_acm.c
Comment (tore-espressif) [inline usb/esp_tinyusb/tusb_cdc_acm.c:412]: Should we unregister the callbacks too?

File: usb/esp_tinyusb/tusb_cdc_acm.c
Comment (tore-espressif) [inline usb/esp_tinyusb/tusb_cdc_acm.c:411]: Could you please review the `free_obj` and `deinit` functions? I find it quite confusing. Could we merge the allocation and initialization (and analogously freeing and deinitialization) to just one function?

**Review by tore-espressif — COMMENTED** [review]
Could you also please bump version of `esp_tinyusb/idf_component.yml` to `1.0.3`?

This way, the component will be uploaded to IDF component registry, after merging

File: usb/esp_tinyusb/tusb_cdc_acm.c
Comment (esp-saurabhbansal) [reply usb/esp_tinyusb/tusb_cdc_acm.c]: done

File: usb/esp_tinyusb/tusb_cdc_acm.c
Comment (esp-saurabhbansal) [reply usb/esp_tinyusb/tusb_cdc_acm.c:401]: No need.
On registering callbacks, these are assigned in CDC_ACM object esp_tusb_cdcacm_t. 
free_obj(itf); actually frees up the memory allocated by esp_tusb_cdcacm_t struct. So, unregistering of callbacks are not needed because memory is getting freed up.

File: usb/esp_tinyusb/tusb_cdc_acm.c
Comment (esp-saurabhbansal) [reply usb/esp_tinyusb/tusb_cdc_acm.c:400]: esp_tusb_cdcacm_t is a subclass specific object (i.e. ACM) which is available only in tusb_sdc_acm.c. It is hidden from generic cdc class source file cdc.c.
alloc_obj()/free_obj() is for allocation/freeing the subclass specific object esp_tusb_cdcacm_t (Here, only ACM type) and thus used only in tusb_sdc_acm.c. By this design and segregation, we should not move the subclass specific data to generic cdc class file.

Comment (esp-saurabhbansal) [reply usb/esp_tinyusb/tusb_cdc_acm.c:400]: @tore-espressif Please check and approve if no more comments.

Comment (tore-espressif) [reply usb/esp_tinyusb/tusb_cdc_acm.c:400]: OK, agreed

**Review by tore-espressif — APPROVED** [review]

File: usb/esp_tinyusb/tusb_cdc_acm.c
Comment (roma-jam) [inline usb/esp_tinyusb/tusb_cdc_acm.c:395]: Hi Saurabh, 
Great job!

One thing that I want to clarify for me is the next. During the tusb_cdc_acm_init function we:
- init tinyusb class device (with tinyusb_cdc_init())
- allocate cdc acm object
- create mutex (ringbuf_read_mux)
- allocate rx_tfbuf 
- allocate ring buffer (rx_unread_buf). 

During my attempts to understand the logic I see, that two main things should be achieved at the end:
1. We can use label goto when we need to separate some different steps from each other.
2. We should free all object, that was allocated at the step/steps before. 

In this case I am concern about the logic, when we need to use only one label (As much as much often at this time we can avoid using that at all). 
The second one tells us that we need to get freed all object had been allocated. That means that we need to do it in a backwards order: 
- free ring buffer
- free rx buffer
- delete mutex
- free cdc acm object
- deinit class usb device

From that point of view, it looks that with the goto labels we can do the following: 
`

    init_tinyusb_class_device();
        if not goto init_fail;
    allocate_cdc_acm_object();
        if not goto allocate_object_fail;
    create_mutex();
        if not goto create_mutex_failure;
    allocate_rx_tfbuf();
        if not goto alloc_buf_err;
    allocate_ring_buffer();
         if not goto ring_buffer_fail;
    return OK;

    ring_buffer_fail:
         free_rx_buffer();
    alloc_buf_error:
         delete_mutex();
    create_mutex_failure:
        free_cdc_acm_object();
    allocate_object_fail:
        deinit_class_usb_device();
    init_fail: 
        // nothing else to free
        return ERR;
`

Honestly, I am not a fan of the goto labels, but in such case it seems to be at least justified. 
Maybe I not understand the logic or I do not have the whole image...But it looks that in the suggested logic we: 
- use goto labels for their purpose
- freed all the object had been allocated... 

That is the only thing I would like to highlight.

Comment (tore-espressif) [reply usb/esp_tinyusb/tusb_cdc_acm.c:395]: C standard says that passing NULL to `free` does nothing. So in order to have the clean-up procedure as trivial as possible, we usually use only one label where we free everything (even if the allocation did not occur). Does that make sense to you @roma-jam ?

Very good answer for this question, for you reference: https://stackoverflow.com/questions/6084218/is-it-good-practice-to-free-a-null-pointer-in-c

Comment (roma-jam) [reply usb/esp_tinyusb/tusb_cdc_acm.c:395]: @tore-espressif If the clean-up procedure is only for freeing allocate memory - yes, that does make perfect sense.

Comment (tore-espressif) [reply usb/esp_tinyusb/tusb_cdc_acm.c:395]: Are you OK with merging this?

Comment (roma-jam) [reply usb/esp_tinyusb/tusb_cdc_acm.c:395]: @tore-espressif I am not sure. 
Probably we need to delete mutex acm->ringbuf_read_mux in case acm->rx_tfbuf allocation error. 
I do not see it in code.

Comment (esp-saurabhbansal) [reply usb/esp_tinyusb/tusb_cdc_acm.c]: @roma-jam @tore-espressif Please check again the new code with additional goto labels.

**Review by roma-jam — APPROVED** [review]
LGTM

**Pull request merged**
